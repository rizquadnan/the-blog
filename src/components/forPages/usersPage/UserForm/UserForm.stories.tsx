import * as React from 'react'
import { Meta, Story } from '@storybook/react'
import { UserForm } from './UserForm'

type TUserForm = React.ComponentProps<typeof UserForm>
export default {
  title: 'User Inputs/UserForm',
  component: UserForm,
} as Meta<TUserForm>

const Template: Story<TUserForm> = (args) => <UserForm {...args} />

export const Create = Template.bind({})
Create.args = {
  variant: 'create',
}

export const Update = Template.bind({})
Update.args = {
  variant: 'update',
  initialValues: {
    email: 'Some email',
    gender: 'female',
    status: 'active',
    name: 'Some name',
  },
}
