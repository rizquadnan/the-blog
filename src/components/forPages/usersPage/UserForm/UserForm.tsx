import { Input } from '@chakra-ui/input'
import { Button } from '@chakra-ui/button'
import {
  FormControl,
  FormErrorMessage,
  FormLabel,
} from '@chakra-ui/form-control'
import React from 'react'
import { Controller, useForm } from 'react-hook-form'
import { VStack } from '@/components/uiKit/VStack'
import { HStack, Radio, RadioGroup } from '@chakra-ui/react'

type TFormValues = {
  name: string
  gender: 'male' | 'female'
  email: string
  status: 'active' | 'inactive'
}
type TUserFormCreate = {
  variant: 'create'
  onSubmit: (props: TFormValues) => void
}
type TUserFormUpdate = {
  variant: 'update'
  initialValues: TFormValues
  onSubmit: (props: TFormValues) => void
}

export function UserForm(props: TUserFormCreate | TUserFormUpdate) {
  const {
    register,
    handleSubmit,
    formState: { errors, isDirty },
    control,
  } = useForm<TFormValues>({
    defaultValues:
      props.variant === 'update'
        ? props.initialValues
        : { email: '', gender: 'male', name: '', status: 'inactive' },
  })

  const notEligbleToUpdate = props.variant === 'update' && !isDirty

  return (
    <VStack as="form" minW="280px">
      <FormControl isInvalid={Boolean(errors.email)}>
        <FormLabel>Email</FormLabel>
        <Input
          {...register('email', {
            required: 'Email is required',
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: 'Invalid email address',
            },
          })}
          placeholder="Email"
          type="email"
        />
        <FormErrorMessage>{errors.email?.message}</FormErrorMessage>
      </FormControl>

      <FormControl isInvalid={Boolean(errors.name)}>
        <FormLabel>Name</FormLabel>
        <Input
          {...register('name', {
            required: 'Name is required',
            minLength: {
              value: 3,
              message: 'Name must be at least 3 characters',
            },
          })}
          placeholder="Name"
        />
        <FormErrorMessage>{errors.name?.message}</FormErrorMessage>
      </FormControl>

      <Controller
        name="gender"
        control={control}
        render={({ field: { onChange, value } }) => (
          <FormControl as="fieldset">
            <FormLabel as="legend">Gender</FormLabel>
            <RadioGroup onChange={onChange} value={value}>
              <HStack spacing="4px">
                <Radio value="male">Male</Radio>
                <Radio value="female">Female</Radio>
              </HStack>
            </RadioGroup>
          </FormControl>
        )}
      />

      <Controller
        name="status"
        control={control}
        render={({ field: { onChange, value } }) => (
          <FormControl as="fieldset">
            <FormLabel as="legend">Status</FormLabel>
            <RadioGroup onChange={onChange} value={value}>
              <HStack spacing="4px">
                <Radio value="active">Active</Radio>
                <Radio value="inactive">Inactive</Radio>
              </HStack>
            </RadioGroup>
          </FormControl>
        )}
      />

      <Button
        w="100%"
        mt="48px !important"
        onClick={handleSubmit(props.onSubmit)}
        isDisabled={notEligbleToUpdate}
      >
        Submit
      </Button>
    </VStack>
  )
}

export default UserForm
