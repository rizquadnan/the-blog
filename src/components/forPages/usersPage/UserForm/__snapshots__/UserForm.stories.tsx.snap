// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Inputs/UserForm Create smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Email",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Email",
      "pressed": undefined,
      "role": "textbox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Name",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Name",
      "pressed": undefined,
      "role": "textbox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Gender",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": true,
      "children": undefined,
      "name": "Male",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Male",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": false,
      "children": undefined,
      "name": "Female",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Female",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Status",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": false,
      "children": undefined,
      "name": "Active",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Active",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": true,
      "children": undefined,
      "name": "Inactive",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Inactive",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Submit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "user-inputs-userform--create",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`User Inputs/UserForm Update smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Email",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "invalid": "true",
      "name": "Email",
      "pressed": undefined,
      "role": "textbox",
      "value": "Some email",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Name",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Name",
      "pressed": undefined,
      "role": "textbox",
      "value": "Some name",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Gender",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": false,
      "children": undefined,
      "name": "Male",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Male",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": true,
      "children": undefined,
      "name": "Female",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Female",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Status",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": true,
      "children": undefined,
      "name": "Active",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Active",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": false,
      "children": undefined,
      "name": "Inactive",
      "pressed": undefined,
      "role": "radio",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Inactive",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "disabled": true,
      "name": "Submit",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "user-inputs-userform--update",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
